File: core/02-streaming/python/tools.py
"""tools.py —— 工具注册表，沿用 01 demo 的写法。
这里只放一个 `get_weather`，演示流式 + 工具调用的 delta 累积就够了。"""
from __future__ import annotations

import json
from typing import Callable

TOOLS: dict[str, tuple[Callable, dict]] = {}


def tool(schema: dict):
    def deco(fn: Callable) -> Callable:
        TOOLS[schema["name"]] = (fn, schema)
        return fn
    return deco


@tool({
    "name": "get_weather",
    "description": "获取指定城市的天气",
    "parameters": {
        "type": "object",
        "properties": {"city": {"type": "string"}},
        "required": ["city"],
    },
})
def get_weather(city: str) -> dict:
    db = {"北京": (15, "晴"), "上海": (20, "多云"), "深圳": (25, "小雨")}
    t, cond = db.get(city, (18, "数据不可用"))
    return {"city": city, "temperature": t, "condition": cond}


def schemas() -> list[dict]:
    return [{"type": "function", "function": s} for _, s in TOOLS.values()]


def call(name: str, args: dict) -> str:
    fn, _ = TOOLS.get(name, (None, None))
    if fn is None:
        return json.dumps({"error": f"unknown tool: {name}"}, ensure_ascii=False)
    try:
        return json.dumps(fn(**args), ensure_ascii=False)
    except Exception as e:
        return json.dumps({"error": str(e)}, ensure_ascii=False)

File: core/02-streaming/go/tools.go
// tools.go —— 工具注册表，沿用 01 demo 的写法。
// 这里只放一个 get_weather，演示流式 + 工具的 delta 累积就够了。
package main

import (
	"encoding/json"
	"fmt"

	"github.com/sashabaranov/go-openai"
	"github.com/sashabaranov/go-openai/jsonschema"
)

type toolFn func(args map[string]any) any

type tool struct {
	def openai.Tool
	fn  toolFn
}

var registry = map[string]tool{}

func register(name, desc string, params jsonschema.Definition, fn toolFn) {
	registry[name] = tool{
		def: openai.Tool{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name: name, Description: desc, Parameters: params,
			},
		},
		fn: fn,
	}
}

func schemas() []openai.Tool {
	out := make([]openai.Tool, 0, len(registry))
	for _, t := range registry {
		out = append(out, t.def)
	}
	return out
}

func call(name, argsJSON string) string {
	t, ok := registry[name]
	if !ok {
		return errJSON(fmt.Sprintf("unknown tool: %s", name))
	}
	var args map[string]any
	if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
		return errJSON(err.Error())
	}
	b, _ := json.Marshal(t.fn(args))
	return string(b)
}

func errJSON(msg string) string {
	b, _ := json.Marshal(map[string]string{"error": msg})
	return string(b)
}

func init() {
	register("get_weather", "获取指定城市的天气",
		jsonschema.Definition{
			Type: jsonschema.Object,
			Properties: map[string]jsonschema.Definition{
				"city": {Type: jsonschema.String, Description: "城市名"},
			},
			Required: []string{"city"},
		},
		func(args map[string]any) any {
			db := map[string]struct {
				t    int
				cond string
			}{"北京": {15, "晴"}, "上海": {20, "多云"}, "深圳": {25, "小雨"}}
			city, _ := args["city"].(string)
			w, ok := db[city]
			if !ok {
				w.t, w.cond = 18, "数据不可用"
			}
			return map[string]any{"city": city, "temperature": w.t, "condition": w.cond}
		},
	)
}
